### DISCUSSION TITLE
Correlating requests with a method invoked later in the call stack for new integration

### DISCUSSION THREAD

**Discussion by rombert** [open-telemetry/opentelemetry-java-instrumentation#9384] [Q&A] [answered]
Hi,

I am working on adding support for [Apache Sling](https://sling.apache.org/) to Open Telemetry. Sling is a Java-based web application framework, and the first item on my list is to customize the span name to reflect the handler.

Sling embeds Jetty has a well-defined entry point in the so-called SlingMainServlet, so what OTEL initially seems is a mapping for `/*`.

When Sling handlers a request it has a `service` entry point method, but associating a request with a "handler" (servlet resolution in Sling jargon → ServletResolver ) happens much later in the stack. I can't "just" instrument the servlet resolution because when handling a request in Sling there is usually a top-level servlet and multiple 'child' servlets which compose the result.

I have two rough ideas right now:

- "mark" the top-level request as it arrives in the SlingMainServlet and instrument the span name during servlet resolution if the request is a top-level one. I _could_ set a well-known request attribute and read it down the line.
- use a thread-local in the ServletResolver to see if it has been invoked already ; but since I don't have a good way of cleaning up without tracking request lifecycle

Does any of these ideas sound workable form an OTEL point of view? Are there any other patters in the java library that could be useful in my scenario?

Thanks!

**Comment by laurit** [comment, marked as answer]
When you just want to update the `server` span name then usually we use https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/db7eb14384bc3d91ebdeaf7c90b1bb0cdb15ba03/instrumentation/play/play-mvc/play-mvc-2.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/play/v2_6/Play26Singletons.java#L63 Set route source to `CONTROLLER` to overwrite the value from the servlet instrumentation or `NESTED_CONTROLLER` like in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/db7eb14384bc3d91ebdeaf7c90b1bb0cdb15ba03/instrumentation/jaxrs/jaxrs-2.0/jaxrs-2.0-jersey-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrs/v2_0/JerseySpanName.java#L28 to overwrite all previous values.

**Reply by rombert** [reply]
Thanks @laurit . By briefly looking at the code and also via testing, it seems that when I use the following snippet the first call for a given span is used for the update and the rest are ignored.

```java
HttpServerRoute.update(context, HttpServerRouteSource.CONTROLLER, route); 
```

Is that correct? If it is, it would greatly simplify the work I would need to do.

**Reply by trask** [reply]
> it seems that when I use the following snippet the first call for a given span is used for the update and the rest are ignored

yes, the HttpServerRoute only updates the span name and `http.route` attribute the top-most SERVER span

**Reply by rombert** [reply]
Great, thanks for confirming @trask .
